cosmwasm/contracts/ibc-translator/src/reply.rs

```diff
@@ -0,0 +1,174 @@
+                chain,
+                Some(payload),
+            )
+        }
+    }
+}
+pub fn convert_cw20_to_bank_and_send(
+    deps: DepsMut,
+    env: Env,
+    recipient: String,
+    amount: u128,
+    cw20_contract_addr: String,
+    chain_id: u16,
+    payload: Option<Binary>,
+) -> Result<Response<TokenFactoryMsg>, anyhow::Error> {
+    deps.api
+        .addr_validate(&cw20_contract_addr)
+        .context(format!("invalid contract address {cw20_contract_addr}"))?;
+
+    // convert contract address into base64
+    let subdenom = contract_addr_to_base58(deps.as_ref(), cw20_contract_addr.clone())?;
+    // format the token factory denom
+    let tokenfactory_denom = "factory/".to_string()
+        + env.contract.address.to_string().as_ref()
+        + "/"
+        + subdenom.as_ref();
+
+    let mut response: Response<TokenFactoryMsg> = Response::new();
+
+    // check contract storage see if we've created a denom for this cw20 token yet
+    // if we haven't created the denom, then create the denom
+    // info.sender contains the cw20 contract address
+    if !CW_DENOMS.has(deps.storage, cw20_contract_addr.clone()) {
+        // call into token factory to create the denom
+        let create_denom = SubMsg::new(TokenMsg::CreateDenom {
+            subdenom,
+            metadata: None,
+        });
+        response = response.add_submessage(create_denom);
+
+        // add the contract_addr => tokenfactory denom to storage
+        CW_DENOMS
+            .save(deps.storage, cw20_contract_addr, &tokenfactory_denom)
+            .context("failed to save contract_addr => tokenfactory denom to storage")?;
+    }
+
+    // add calls to mint and send bank tokens
+    response = response.add_message(TokenMsg::MintTokens {
+        denom: tokenfactory_denom.clone(),
+        amount,
+        mint_to_address: env.contract.address.to_string(),
+    });
+
+    let channel = CHAIN_TO_CHANNEL_MAP
+        .load(deps.storage, chain_id)
+        .context("chain id does not have an allowed channel")?;
+
+    let payload_decoded = match payload {
+        Some(payload) => String::from_utf8(payload.to_vec())
+            .context(format!("failed to convert {payload} to utf8 string"))?,
+        None => "".to_string(),
+    };
+
+    // Create MsgTransfer protobuf message for Stargate
+    // https://github.com/cosmos/ibc-go/blob/main/proto/ibc/applications/transfer/v1/tx.proto#L27
+    let ibc_msg_transfer = Anybuf::new()
+        .append_string(1, "transfer") // source port
+        .append_string(2, channel) // source channel
+        .append_message(
+            3,
+            &Anybuf::new()
+                .append_string(1, tokenfactory_denom)
+                .append_string(2, amount.to_string()),
+        ) // Token
+        .append_string(4, env.contract.address) // sender
+        .append_string(5, recipient) // receiver
+        .append_message(6, &Anybuf::new().append_uint64(1, 0).append_uint64(2, 0)) // TimeoutHeight
```

Comment:
How is the timeout height determined? Is this according to a spec somewhere?

Reply:
This line of code should be configuring the `TimeoutHeight` to 0. `TimeoutHeight` should be 0 since we are using the `TimeoutTimestamp` to specify timeouts.

@misko9 to confirm

Reply:
That is correct, TimeoutHeight is set to 0 to use TimeoutTimestamp.